tools/chapel-py/chapel.cpp

```diff
@@ -0,0 +1,751 @@
+  PyObject* topExprs = PyList_New(listSize);
+  for (auto i = 0; i < listSize; i++) {
+    PyObject* node = wrapAstNode(self, builderResult.topLevelExpression(i));
+    PyList_SetItem(topExprs, i, node);
+  }
+  return topExprs;
+}
+
+static PyObject* ContextObject_is_bundled_path(ContextObject *self, PyObject* args) {
+  auto context = &self->context;
+  const char* fileName;
+  if (!PyArg_ParseTuple(args, "s", &fileName)) {
+    PyErr_BadArgument();
+    return nullptr;
+  }
+  auto pathUS = chpl::UniqueString::get(context, fileName);
+
+  bool isInternalPath =
+    chpl::parsing::filePathIsInInternalModule(context, pathUS) ||
+    chpl::parsing::filePathIsInStandardModule(context, pathUS) ||
+    chpl::parsing::filePathIsInBundledModule(context, pathUS);
+
+  return PyBool_FromLong(isInternalPath);
+}
+
+static PyMethodDef ContextObject_methods[] = {
+  { "parse", (PyCFunction) ContextObject_parse, METH_VARARGS, "Parse a top-level AST node from the given file" },
+  { "is_bundled_path", (PyCFunction) ContextObject_is_bundled_path, METH_VARARGS, "Check if the given file path is within the bundled (built-in) Chapel files" },
+  {NULL, NULL, 0, NULL}  /* Sentinel */
+};
+
+PyTypeObject ContextType = {
+  PyVarObject_HEAD_INIT(NULL, 0)
+  .tp_name = "Context",
+  .tp_basicsize = sizeof(ContextObject),
+  .tp_itemsize = 0,
+  .tp_dealloc = (destructor) ContextObject_dealloc,
+  .tp_flags = Py_TPFLAGS_DEFAULT,
+  .tp_doc = PyDoc_STR("The Chapel context object that tracks various frontend state"),
+  .tp_methods = ContextObject_methods,
+  .tp_init = (initproc) ContextObject_init,
+  .tp_new = PyType_GenericNew,
+};
+
+static int AstNodeObject_init(AstNodeObject* self, PyObject* args, PyObject* kwargs) { \
+  PyObject* contextObjectPy;
+  if (!PyArg_ParseTuple(args, "O", &contextObjectPy))
+      return -1;
+
+  Py_INCREF(contextObjectPy);
+  self->astNode = nullptr;
+  self->contextObject = contextObjectPy;
+  return 0;
+}
+
+static void AstNodeObject_dealloc(AstNodeObject* self) {
+  Py_XDECREF(self->contextObject);
+  Py_TYPE(self)->tp_free((PyObject *) self);
+}
+
+static PyObject* AstNodeObject_dump(AstNodeObject *self, PyObject *Py_UNUSED(ignored)) {
+  self->astNode->dump();
+  Py_RETURN_NONE;
+}
+
+static PyObject* AstNodeObject_tag(AstNodeObject *self, PyObject *Py_UNUSED(ignored)) {
+  const char* nodeType = chpl::uast::asttags::tagToString(self->astNode->tag());
+  return Py_BuildValue("s", nodeType);
+}
+
+static PyObject* AstNodeObject_unique_id(AstNodeObject *self, PyObject *Py_UNUSED(ignored)) {
+  auto uniqueID = (intptr_t)self->astNode;
+  return Py_BuildValue("K", uniqueID);
+}
+
+
+static PyObject* AstNodeObject_attribute_group(AstNodeObject *self, PyObject *Py_UNUSED(ignored)) {
+  auto attributeGroup = self->astNode->attributeGroup();
+  if (attributeGroup != nullptr) {
+    return wrapAstNode((ContextObject*) self->contextObject, attributeGroup);
+  }
+  Py_RETURN_NONE;
+}
```

Comment:
I don't think this is needed? I think you can just do
```c++
auto attributeGroup = self->astNode->attributeGroup();
return wrapAstNode((ContextObject*) self->contextObject, attributeGroup);
```
Since `wrapAstNode` handles the nullptr.